tetj/tetj.3
.TH tetj "3" "" "tetj (LSB)" LSB
.SH NAME
tetj \- TET journal emulation toolkit
.SH DESCRIPTION
.PP
The
.I tetj
library provides a simple toolkit to generate a 
.I journal
file in the format produced by 
.IR TET , 
the Test Environment Toolkit.
Many LSB Certification tests are written to use
TETware-lite as a test harness, and thus certification
depends on results being available in this format.
For tests which do not use the TET framework,
tetj is a way to generate a compatible journal
which can be analyzed by existing tools.
.PP
The general structure of the journal file is some initial
setup information, followed by structured information in
begin/end pairs, wrapping other begin/end pairs which
contain result and informational messages.
Each line in a journal file consists of three fields
separated by | characters.  The first field contains
a number which indicates the type of the journal line.
The second field contains information depending on the
type of the journal line.
The third field generally contains some kind of message text. 
.PP
.I tetj
provides functions to generate the most common journal
line types, and can easily be extended to generate
additional line types if required.
.SH SYNOPSIS
.nf
.B #include "tetj.h"

.B cc files libtetj.a
.fi
.PP
Programs should initialize use of the library by calling the
.I tetj_start_journal
function with: a filename for the journal; a 
.B tetj_handle 
pointer, for which the function will allocate storage space
and fill in the pointer; and the command line the test was called
with.  The command line and some collected
system information will be recorded at the start of the
journal file, which will thus have two lines in it on
return.  The handle to the journal must be kept by the test suite
and passed to all of the other library routines as the first argument.
.PP
A journal file is completed by a call to 
.IR tetj_close_journal ,
which writes ending information and closes the journal file;
this call is mandatory as the journal will be considered
invalid if the ending information is absent.
.PP
Any configuration information describing parameters
that affect the execution of the test is recorded with
one or more calls to the 
.I tetj_add_config
function;  this block of header information should
be preceded by a call to the
.I tetj_config_start
function and followed by a call to the
.I tetj_config_end
function.
.PP
If the test suite knows how many tests will be executed,
this information should be recorded with an optional call to the
.I tetj_scenario_info
function, the format of the 
.B message 
argument is, assuming test "foo" and a count
of "27" (the quotes are required):
.nf
\fB"total tests in foo 27"\fR
.fi
.PP
The remaining contents of the journal file is made up of
information about tests executed during the test run.
Tests can be single or dual level; the top level is
the "test case", the second level is the "test purpose".
There may be one or more test cases, which are set up
and terminated by calls to
.I tetj_testcase_start
and
.IR tetj_testcase_end ,
and each test case may have one or more test purposes,
which are set up and terminated by calls to
.I tetj_purpose_start
and
.IR tetj_purpose_end .
Each test purpose must record a result; the journal
is considered invalid if it is lacking any expected
result code. These are recorded by a call to
.IR tetj_result .
The result code is selected from the predefined
values in the 
.B testcase_results
enum, these codes have specific meanings and should
not be altered (see 
.IR tetj.h ).
If it would be helpful to record more information about
a test, perhaps the reason why a failure is being
reported, this can be done with one or more calls to
.IR tetj_testcase_info ,
which should precede the
.I tetj_result
call.
.PP
The test suite is responsible for keeping track of two counters,
.B tetj_activity_count
and
.BR tetj_tp_count ,
both of which are declared and initialized in
.IR tetj.h .
.B tetj_activity_count
(starts at 0) should be incremented for each new testcase while
.B tetj_tp_count 
should be reset for each new testcase (starts at 1) 
and incremented for each test purpose. 
.B tetj_activity_count 
is passed to each function that takes an "activity" and 
.B tetj_tp_count 
to each function that takes a "tpnumber" argument.
The intent is that each test purpose be uniquely identified
by a pair of serial numbers; the journal will be considered
invalid if this is not true.
Note: since these are global variables, it seems strange
that they must be explicitly passed - perhaps a future
enhancement may remove this requirement.
.PP
As a convenience, an additional global variable
.B tetj_arch
records the machine architecture. 
This information is captured during the execution of
.I tetj_start_journal
and saved, since it is needed by test programs which
are to be used for LSB certification in constructing
a configuration information line.  The standard format
for this line, which is written with a call to
.IR tetj_add_config ,
is:

	30||VSX_NAME=\fItestid\fR \fIversion\fR (\fIarch\fR)

When a line of this format is detected, the report generator
.B tjreport
will attempt to obtain an official waiver file matching
the (testid, version, arch) triple.
.PP
The 
.B testcase
name passed to
.I tetj_testcase_start
is required to be unique across the entire journal.
When tests are executed using the TET toolkit,
these testcase names map to entries in the scenario
file, thus the reporting tools consider a journal
invalid if the testcase names are not unique.
.PP
.nf
struct tetj_handle;
extern int tetj_activity_count;
extern int tetj_tp_count;
extern char *tetj_arch;
enum testcase_results { ... };

int tetj_start_journal(const char *pathname, struct tetj_handle **handle,
                       char *command_run);
int tetj_close_journal(struct tetj_handle *handle);
void tetj_config_start(struct tetj_handle *handle);
void tetj_add_config(struct tetj_handle *handle, char *message);
void tetj_config_end(struct tetj_handle *handle);
void tetj_scenario_info(struct tetj_handle *handle, char *message);
void tetj_add_controller_error(struct tetj_handle *handle,
                               char *message);
void tetj_testcase_start(struct tetj_handle *handle,
                         unsigned int activity, char *testcase,
                         char *message);
void tetj_testcase_end(struct tetj_handle *handle,
                       unsigned int activity, unsigned int status,
                       char *message);
void tetj_purpose_start(struct tetj_handle *handle,
                        unsigned int activity, unsigned int tpnumber,
                        char *message);
void tetj_purpose_end(struct tetj_handle *handle, unsigned int activity,
                      unsigned int tpnumber);
void tetj_result(struct tetj_handle *handle, unsigned int activity,
                 unsigned int tpnumber, enum testcase_results result);
void tetj_testcase_info(struct tetj_handle *handle,
                        unsigned int activity, unsigned int tpnumber,
                        unsigned int context, unsigned int block,
                        unsigned int sequence, char *message);
.fi

.SH "AUTHORS"
The contributors to the Linux Standard Base.
.SH "REPORTING BUGS"
If you obtained this library from the LSB ftp site,
report bugs at http://bugs.linuxbase.org or email to
<lsb-discuss@linux-foundation.org>.  If you obtained this
from your distribution vendor, report bugs back to the
vendor in the normal way.
.SH "SEE ALSO"
tjreport(1)
.br
Linux Standard Base specification and other documents at
http://www.linux-foundation.org/
.PP
The TETware User Guide from The Open Group contains much
more detail on the journal file entries.


ChangeLog
2006-11-20	Jeff Licquia <jeff_at_licquia.org>
	* appchk/output.c
	  appchk/output.h
	  elfchk/appchk_untetj.h
	  appchk/appchk.c
	  appchk/check_file.c
	  appchk/check_file.h
	  appchk/makefile
	  appchk/symbols.c
	  appchk/symbols.h
	  archk/archk.c
	  archk/makefile
	  elfchk/dynamic.c
	  elfchk/makefile
	  elfchk/progbits.c
	  elfchk/proghdr.c
	  pkgchk/pkgchk.c
	  rpmchk/archive.c
	   Create new report output, and make it default.
	   Support --shared-libpath (add all libs in dir).
2006-07-27	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* elfchk/progbits.c
	  elfchk/dwarf.c
	  elfchk/dwarf.h
	  elfchk/progbits.c
	   don't fall off end of eh_frame section (bug 1444)

2006-04-26	Marvin Heffler <heffler_at_us.ibm.com>
	* rpmchk/vals.c
	   remove lsb-desktop-noarch as a valid dependency and add
	   lsb-graphics-<arch> as an optional dependency

2006-04-18	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* libchk/libchk.c
	   bug 1241: logic change in case the first instance of a symbol is
	   unversioned but we're expecting versioned - keep looking!

2006-04-10	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* libchk/libchk.c
	   misc. cleanup (no functional changes)

2006-04-04	Marvin Heffler <heffler_at_us.ibm.com>
	* rpmchk/vals.c
	  pkgchk/pkgchk.c
	  rpmchk/dependencies.c
	  rpmchk/rpmchk.h
	  rpmchk/vals.c
	   comment out all is_graphics and is_desktop processing since LSB
	   3.1 defaults to desktop

2006-04-01	Rajesh Banginwar <rajesh.banginwar_at_intel.com>
	* rpmchk/vals.c
	  appchk/appchk.c
	  appchk/lsbappchk.1
	  cmdchk/cmdchk.c
	  cmdchk/lsbcmdchk.1
	  elfchk/dump_modules
	  elfchk/modules.c
	  elfchk/modules.h
	  elfchk/util.c
	  libchk/libchk.c
	  libchk/lsblibchk.1
	  pkgchk/lsbpkgchk.1
	  pkgchk/pkgchk.c
	   undocument -T and make desktop default. clean-up to simplify
	   appchk messages

2006-03-30	Marvin Heffler <heffler_at_us.ibm.com>
	* libchk/libchk.c
	  elfchk/modules.c
	  elfchk/modules.h
	  elfchk/util.c
	   change libchk so it handles module sets better

2006-03-24	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* libchk/dummy_link.c
	   try a qt-mt symbol that isn't arch-specific

2006-03-21	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* libchk/libqt_mt_so_3.c
	   Added missing file for Qt3

2006-03-21	Rajesh Banginwar <rajesh.banginwar_at_intel.com>
	* libchk/makefile
	  elfchk/dtneeded.c
	  elfchk/dynsyms.c
	  elfchk/modules.c
	  elfchk/modules.h
	  libchk/dummy_link.c
	  libchk/libs.c
	  libchk/libs.h
	  libchk/libs.mk
	   sync with DB. enabling for Qt3

2006-03-15	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* libchk/libstdcxx_so_6.c
	  devchk/ts/devchk/glib_2_0_glib_h.c
	  elfchk/dtneeded.c
	  libchk/libQtGui_so_4.c
	  libchk/libgdk_pixbuf_xlib_2_0_so_0.c
	  libchk/libs.c
	  libchk/libs.h
	  libchk/libs.mk
	   regen for bug 1288

2006-03-13	Marvin Heffler <heffler_at_us.ibm.com>
	* devchk/ts/devchk/sys_shm_h.c
	  devchk/ts/devchk/png_h.c
	   restore hand edit workaround to get png_h.c to compile with
	   lsbcc

2006-03-13	Mats Wichmann <mats.d.wichmann_at_intel.com>
	* devchk/ts/devchk/wordexp_h.c
	  devchk/ts/devchk/atk_1_0_atk_atk_h.c
	  devchk/ts/devchk/ctype_h.c
	  devchk/ts/devchk/fontconfig_fontconfig_h.c
	  devchk/ts/devchk/ftw_h.c
	  devchk/ts/devchk/glib_2_0_glib_h.c
	  devchk/ts/devchk/glib_2_0_glib_object_h.c
	  devchk/ts/devchk/glib_2_0_gmodule_h.c
	  devchk/ts/devchk/glib_2_0_gobject_gvaluecollector_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_gdk_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_gdk_pixbuf_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_gdk_pixdata_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_xlib_gdk_pixbuf_xlib_h.c
	  devchk/ts/devchk/gtk_2_0_gtk_gtk_h.c
	  devchk/ts/devchk/hdrchk.h
	  devchk/ts/devchk/jpeglib_h.c
	  devchk/ts/devchk/libxml2_libxml_HTMLparser_h.c
	  devchk/ts/devchk/libxml2_libxml_SAX2_h.c
	  devchk/ts/devchk/libxml2_libxml_catalog_h.c
	  devchk/ts/devchk/libxml2_libxml_encoding_h.c
	  devchk/ts/devchk/libxml2_libxml_entities_h.c
	  devchk/ts/devchk/libxml2_libxml_parser_h.c
	  devchk/ts/devchk/libxml2_libxml_pattern_h.c
	  devchk/ts/devchk/libxml2_libxml_relaxng_h.c
	  devchk/ts/devchk/libxml2_libxml_schematron_h.c
	  devchk/ts/devchk/libxml2_libxml_tree_h.c
	  devchk/ts/devchk/libxml2_libxml_valid_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlIO_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlerror_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlmodule_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlreader_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlregexp_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlsave_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlschemas_h.c
	  devchk/ts/devchk/libxml2_libxml_xmlschemastypes_h.c
	  devchk/ts/devchk/libxml2_libxml_xpath_h.c
	  devchk/ts/devchk/makefile
	  devchk/ts/devchk/mktests
	  devchk/ts/devchk/pango_1_0_pango_pango_h.c
	  devchk/ts/devchk/png_h.c
	  devchk/ts/devchk/regex_h.c
	  devchk/ts/devchk/rpc_auth_h.c
	  devchk/ts/devchk/rpc_clnt_h.c
	  devchk/ts/devchk/rpc_rpc_msg_h.c
	  devchk/ts/devchk/rpc_svc_h.c
	  devchk/ts/devchk/rpc_xdr_h.c
	  devchk/ts/devchk/sys_resource_h.c
	  devchk/ts/devchk/sys_shm_h.c
	   sync after mktests changes for bug 1277,1278

2006-03-09	Brian Harring <brianx.d.harring_at_intel.com>
	* elfchk/util.c
	   fixed potential segfault; don't deref the ptr if testing it to
	   see if it's null.

2006-03-09	Marvin Heffler <heffler_at_us.ibm.com>
	* libchk/libstdcxx_so_6.c
	  devchk/ts/devchk/atk_1_0_atk_atk_h.c
	  devchk/ts/devchk/fontconfig_fontconfig_h.c
	  devchk/ts/devchk/glib_2_0_glib_h.c
	  devchk/ts/devchk/glib_2_0_glib_object_h.c
	  devchk/ts/devchk/glib_2_0_gmodule_h.c
	  devchk/ts/devchk/glib_2_0_gobject_gvaluecollector_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_gdk_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_gdk_pixbuf_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_gdk_pixdata_h.c
	  devchk/ts/devchk/gtk_2_0_gdk_pixbuf_xlib_gdk_pixbuf_xlib_h.c
	  devchk/ts/devchk/gtk_2_0_gtk_gtk_h.c
	  devchk/ts/devchk/jpeglib_h.c
	  devchk/ts/devchk/libxml2_libxml_HTMLparser_h.c

tetj/tetj.c
/* Interface for generating TET like report journals
 *
 * This intended to be a simple way of creating journals which
 * can be analysed using standard TET journal tools without
 * having to compile or link against the TET libraries.
 * 
 * (C) Copyright 2002 The Free Standards Group Inc
 *
 * 2002/03/19 Chris Yeoh, IBM
 *
 * This is $Revision: 1.7 $
 * 
 * $Log: tetj.c,v $
 * Revision 1.7  2006/01/14 14:09:33  mats
 * Provide support for VSX_NAME standard (bug 986)
 *
 * Revision 1.6  2005/12/15 15:07:05  mats
 * Assorted cleanups: tetjtest and python self test pretty much match now
 * in journal format (python selftest presents tests in different order)
 *
 * Revision 1.5  2005/06/10 18:17:20  mats
 * Add support for 40 (config end) and 70 (scenario info) lines.
 * Update README to reflect what's supported and what isn't.
 *
 * Revision 1.4  2005/05/04 00:07:03  mats
 * Some formatting cleanup to make tet journal tools happier on this
 * emulated journal stuff
 *
 * Revision 1.3  2002/04/29 04:39:06  cyeoh
 * Adds support for 'IC Start' and 'IC End' markers in
 * the journal file
 *
 * Revision 1.2  2002/04/29 04:12:05  cyeoh
 * Adds tetj_purpose_end
 * Adds IC Start markers
 *
 * Revision 1.1  2002/03/19 06:07:49  cyeoh
 * Adds simple interface for generating TET journals
 *
 *
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <sys/utsname.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <pwd.h>
#include <errno.h>

#include "tetj.h"

struct tetj_handle
{
  FILE *journal;
};

int tetj_activity_count = 0;
int tetj_tp_count = 0;
char *tetj_vers = "tetj-1.0";
char *tetj_arch = "(noarch)";

static char *get_current_time_string()
{
  static char time_string[20];
  time_t current_time;

  current_time = time(NULL);
  strftime(time_string, 20, "%H:%M:%S", localtime(&current_time));
  return time_string;
}

static char *get_result_string(enum testcase_results result)
{
  switch (result)
  {
  case TETJ_PASS: 
    return "PASS";
  case TETJ_FAIL:
    return "FAIL";
  case TETJ_UNRESOLVED:
    return "UNRESOLVED";
  case TETJ_NOTINUSE:
    return "NOTINUSE";
  case TETJ_UNSUPPORTED:
    return "UNSUPPORTED";
  case TETJ_UNTESTED:
    return "UNTESTED";
  case TETJ_UNINITIATED:
    return "UNITIATED";
  case TETJ_UNREPORTED:
    return "UNREPORTED";
  case TETJ_WARNING:
    return "WARNING";
  case TETJ_FIP:
    return "FIP";
  case TETJ_NOTIMP:
    return "NOTIMP";
  case TETJ_UNAPPROVE:
    return "UNAPPROVE";
  default:
    return "UNKNOWN";
  }
}

int tetj_start_journal(const char *pathname, struct tetj_handle **handle,
                       char *command_run)
{
  static struct utsname uname_info;
  char datetime[20];
  time_t current_time;
  uid_t uid;
  struct passwd *pwent;

  *handle = malloc(sizeof(struct tetj_handle));

  if ( ((*handle)->journal = fopen(pathname, "w")) == NULL ) {
    return errno;
  }
  
  if (uname(&uname_info)!=0) {
    return errno;
  }    
  tetj_arch = uname_info.machine;

  current_time = time(NULL);
  strftime(datetime, 20, "%H:%M:%S %Y%m%d", localtime(&current_time));
  uid = getuid();
  pwent = getpwuid(uid);
  
  fprintf((*handle)->journal, 
          "0|%s %s|User: %s (%i) TCC Start, Command line: %s\n",
          tetj_vers, datetime, pwent ? pwent->pw_name : "", uid, command_run);

  fprintf((*handle)->journal, "5|%s %s %s %s %s|System Information\n",
          uname_info.sysname, uname_info.nodename, uname_info.release,
          uname_info.version, uname_info.machine);
          
  return 0;
}

int tetj_close_journal(struct tetj_handle *handle)
{
  if (handle) {
    fprintf(handle->journal, "900|%s|TCC End\n", get_current_time_string());
    return fclose(handle->journal);
  } else {
    return 0;
  }
}

void tetj_config_start(struct tetj_handle *handle)
{
  if (handle) {
    fprintf(handle->journal, "20||Config Start\n");
  }
}

void tetj_add_config(struct tetj_handle *handle, char *message)
{
  if (handle) {
    fprintf(handle->journal, "30||%s\n", message);
  }
}

void tetj_config_end(struct tetj_handle *handle)
{
  if (handle) {
    fprintf(handle->journal, "40||Config End\n");
  }
}

void tetj_scenario_info(struct tetj_handle *handle, char *message)
{
  if (handle) {
    fprintf(handle->journal, "70||%s\n", message);
  }
}

void tetj_add_controller_error(struct tetj_handle *handle, char *message)
{
  if (handle) {
    fprintf(handle->journal, "50||%s\n", message);
  }
}

void tetj_testcase_start(struct tetj_handle *handle, 
                         unsigned int activity, char *testcase, 
                         char *message)
{
  if (handle) {
    fprintf(handle->journal, "10|%u %s %s|TC Start %s\n",
            activity, testcase, get_current_time_string(), message);
    fprintf(handle->journal, "15|%u %s 1|TCM Start %s\n", activity, 
	    tetj_vers, message);
  }
}

void tetj_testcase_end(struct tetj_handle *handle, 
                       unsigned int activity, unsigned int status,
                       char *message)
{
  if (handle) {
    fprintf(handle->journal, "80|%u %u %s|TC End %s\n",
            activity, status, get_current_time_string(), message);
  }
}

void tetj_purpose_start(struct tetj_handle *handle,
                        unsigned int activity, unsigned int tpnumber, 
                        char *message)
{
  if (handle) {
    fprintf(handle->journal, "400|%u %u 1 %s|IC Start\n",
            activity, tpnumber, get_current_time_string());
    fprintf(handle->journal, "200|%u %u %s|%s\n",
            activity, tpnumber, get_current_time_string(), message);
  }
}

void tetj_purpose_end(struct tetj_handle *handle,
                      unsigned int activity, unsigned int tpnumber)
{
  if (handle) {
    fprintf(handle->journal, "410|%u %u 1 %s|IC End\n",
            activity, tpnumber, get_current_time_string());
  }
}

void tetj_result(struct tetj_handle *handle, 
                 unsigned int activity, unsigned int tpnumber, 
                 enum testcase_results result)
{
  if (handle) {
    fprintf(handle->journal, "220|%u %u %i %s|%s\n",
            activity, tpnumber, result, get_current_time_string(),
            get_result_string(result));
  }
}

void tetj_testcase_info(struct tetj_handle *handle,
                        unsigned int activity, unsigned int tpnumber,
                        unsigned int context, unsigned int block,
                        unsigned int sequence, char *message)
{
  if (handle) {
    fprintf(handle->journal, "520|%u %u %u %u %u|%s\n",
            activity, tpnumber, context, block, sequence, message);
  }
}
